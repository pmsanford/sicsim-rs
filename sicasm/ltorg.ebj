HLTOR  000000000015
T000000180320066F20063F20064869214142437720033F2FFD444546
E000000
