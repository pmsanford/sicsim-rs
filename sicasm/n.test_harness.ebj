HHARN  000200000006
T000200064B2FFD3F2FFD
E000200
