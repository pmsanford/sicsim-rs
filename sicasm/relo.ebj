HRELO  000000000016
T000000120310000D1B20090F20093F2FFD00000A000005
M00000105
E000000
