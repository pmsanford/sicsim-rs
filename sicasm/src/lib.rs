use anyhow::Result;
use pass_one::FirstPass;
use pass_two::pass_two;
use std::fmt::Write;

mod constants;
mod directive;
mod labels;
mod pass_one;
mod pass_two;
mod record;

/// Assemble a SIC program from text.
///
/// # Errors
///
/// If there's an error in the assembly code
pub fn assemble_program(program_text: &str) -> Result<String> {
    Ok(assemble_with_debug(program_text)?.0)
}

pub fn assemble_with_debug(program_text: &str) -> Result<(String, String)> {
    let lines = program_text.lines().collect::<Vec<_>>();
    let pass_one = FirstPass::parse_lines(&lines)?;

    let (records, debug) = pass_two(pass_one)?;

    let mut assembled = String::new();

    for record in records {
        writeln!(&mut assembled, "{}", record)?;
    }

    Ok((assembled, debug.to_string()?))
}
