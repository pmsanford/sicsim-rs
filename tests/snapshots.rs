use sicasm::assemble_program;

#[test]
fn test_ltorg() {
    let program_text = include_str!("../programs/ltorg.sic");
    let assembled = assemble_program(program_text).unwrap();

    insta::assert_yaml_snapshot!(assembled);
}

#[test]
fn test_copy_xe() {
    let program_text = include_str!("../programs/copy.sic");
    let assembled = assemble_program(program_text).unwrap();

    insta::assert_yaml_snapshot!(assembled);
}

#[test]
fn test_org() {
    let program_text = include_str!("../programs/org.sic");
    let assembled = assemble_program(program_text).unwrap();

    insta::assert_yaml_snapshot!(assembled);
}

#[test]
fn test_word_label_vals() {
    let program_text = include_str!("../programs/word_label.sic");
    let assembled = assemble_program(program_text).unwrap();

    insta::assert_yaml_snapshot!(assembled);
}

#[test]
fn test_equ() {
    let program_text = include_str!("../programs/equ.sic");
    let assembled = assemble_program(program_text).unwrap();

    insta::assert_yaml_snapshot!(assembled);
}

#[test]
fn test_relocate() {
    let program_text = include_str!("../programs/relocatable.sic");
    let assembled = assemble_program(program_text).unwrap();

    insta::assert_yaml_snapshot!(assembled);
}
