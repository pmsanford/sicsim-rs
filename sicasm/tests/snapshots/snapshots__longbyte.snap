---
source: tests/snapshots.rs
expression: assembled
---
"HLBYT  000000000038\nT0000001E0320323F2FFD54686973206973207175697465206C6F6E673B206C6F6E67\nT00001E1A6572207468616E206F6E652074657874207265636F7264000002\nE000000\n"

